### ISSUE TITLE
Isolated points missing in pdf

### ISSUE THREAD

**Issue by arndtp** [matplotlib/matplotlib#7293]
Dear developers,
when plotting single points surrounded by np.nan values, points are shown in a regular plot, but are omitted when saved via pdf backend. Minimal example:

```
import numpy as np
import matplotlib.pyplot as plt

a = np.array([1,np.nan,2,np.nan,3,3.5,4])

plt.plot(a, marker='o')
plt.show()
```

generates a perfect plot with the 5 points shown:

![foo1](https://cloud.githubusercontent.com/assets/17202381/19445433/881a8df0-9494-11e6-853f-893cd63f80f7.png)

but directly saving it via pdf backend omitts the values isolated by np.nans:

```
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.backends.backend_pdf as pdf

a = np.array([1,np.nan,2,np.nan,3,3.5,4])

fig1 = plt.figure()
sub  = fig1.add_subplot(1,1,1)
sub.plot(a,marker='o')
pp1  = pdf.PdfPages('foo.pdf')
fig1.savefig(pp1, format='pdf')
pp1.close()
```

like so:

![foo2](https://cloud.githubusercontent.com/assets/17202381/19445554/0130e8d8-9495-11e6-9a31-a6b70719b87d.png)

I would be glad for any help.

System:
Ubuntu 14.02
Python 2.7.6
Numpy 1.11.2
Matplotlib 1.5.3

**Comment by anntzer** [comment]
Confirmed on 2.0b4.  This affects (at least?) all backends that rely on `convert_to_string` (defined in `src/_path.h`): pdf, ps and svg.

**Comment by tacaswell** [comment]
This is a pretty nasty bug and a regression from 1.5.

Suspect we just need to chop the paths up better?

**Comment by gcalmettes** [comment]
Same problem encountered (matplotlib 2.0.0) when plotting data from a`pandas.DataFrame`. Filtering out the `NaN` fixed it. Thanks @tacaswell.

**Comment by WeatherGod** [comment]
This bug sounds familiar. Didn't we have something like this when we messed
around with the line chopping code originally? Something about handling
chopping on nan segments?

**Comment by tacaswell** [comment]
The `covert_to_string` is a bit of a red herring.  in `Line2D` we draw lines in two passes, the first is the _line_ which goes through `convert_to_string` (which is correctly simplifying out the nan ponts).  The markers are drawn in a second step (well, actually 2 steps as we support 2-tone markers, but I digress) with calls to `draw_markers`.  The issue is that in `path.iter_segments` used in `draw_markers` the clipping logic removes any multiple 'moves' in a row (and only keeps the last one).  We need to see those points to draw the markers.

We don't see this problem in Agg because it's `draw_markers` does not have the clipper in it's path proccesing pipeline.  Not sure yet if this is an oversight or we are relying on the Agg rasterer to do the clipping for us.

**Comment by NelleV** [comment]
Fixed in #8297
